### ISSUE TITLE
export `setNumberValue` and `safeNextStep` from useNumberFieldState

### ISSUE THREAD

**Issue by andelkocvjetkovic** [adobe/react-spectrum#2786]
## 🙋 Feature Request

Hello, in the [useNumberField](https://react-spectrum.adobe.com/react-stately/useNumberFieldState.html), would be nice if setNumberValue and safeNextStep are exported from the module.

**Comment by LFDanLu** [comment]
@andelkocvjetkovic Hi, mind expanding on your need for `setNumberValue` and `safeNextStep`?

**Comment by andelkocvjetkovic** [comment]
Hello @LFDanLu, if I have minValue, maxValue and snap value there is no way to prevent value to be snapped to nearest step. Ex I had to build a money field that allows user to input $120.10 but increment and decrement it by $ 100, I found a workaround it, override `onPressStart` and used `props.onChange` to change  the `value` since `setNumberValue` is not exported and had to write same logic for nextStep as it is in the `safeNextStep`.

**Comment by snowystinger** [comment]
I think I understand your use case, you would want a value of $120.10 when incremented to become $220.10?

What we currently do is match native input behavior. See https://jsfiddle.net/mod7pktu/
```<input type="number" step="100" />```

**Comment by andelkocvjetkovic** [comment]
yup, exactly that, also I couldn't just you input $220.10 because there is on onBlur which is going format the input to snap value.

**Comment by snowystinger** [comment]
Ok, so for $120.10 -> $220.10, I don't think we'll support this, it doesn't match native behavior.

However, not being able to type any number into a field with a step, that seems like a bug we should address.

**Comment by andelkocvjetkovic** [comment]
Okay, any chance that exported api of useNumberFieldState will be expended ?

**Comment by snowystinger** [comment]
I will ask the team how they feel about it

**Comment by snowystinger** [comment]
Hey, so we went back through the history of the component. This is actually our intentional behavior. See https://react-spectrum.adobe.com/react-spectrum/NumberField.html#step-values
> If the user types a value that is between two steps and blurs the input, the value will be snapped to the nearest step.

Now the reason this is intentional is because we have matched slider behavior when provided with steps. A slider cannot be in between two steps, it must be snapped. Steps mean something is now only discreet values. If you need to be able to enter arbitrary values, then it's best not to provide a step.

You can find more information on the behavior here https://github.com/adobe/react-spectrum/pull/1509 where it was introduced.

As such, we have no plans right now to change the behavior of step snapping.


We have no plans to export `setNumberValue`, you should be able to accomplish this using the prop `value` passed into the hook.
For `safeNextStep`, we also do not plan to export this either.

Now that's still not to say we wouldn't consider it. However, we need more information. Above you said you came up with a workaround. Could you share that in a codesandbox so we can better understand what you're doing? Maybe we'll have other suggestions.

**Comment by andelkocvjetkovic** [comment]
> Hey, so we went back through the history of the component. This is actually our intentional behavior. See https://react-spectrum.adobe.com/react-spectrum/NumberField.html#step-values
> 
> > If the user types a value that is between two steps and blurs the input, the value will be snapped to the nearest step.
> 
> Now the reason this is intentional is because we have matched slider behavior when provided with steps. A slider cannot be in between two steps, it must be snapped. Steps mean something is now only discreet values. If you need to be able to enter arbitrary values, then it's best not to provide a step.
> 
> You can find more information on the behavior here #1509 where it was introduced.
> 
> As such, we have no plans right now to change the behavior of step snapping.
> 
> We have no plans to export `setNumberValue`, you should be able to accomplish this using the prop `value` passed into the hook. For `safeNextStep`, we also do not plan to export this either.
> 
> Now that's still not to say we wouldn't consider it. However, we need more information. Above you said you came up with a workaround. Could you share that in a codesandbox so we can better understand what you're doing? Maybe we'll have other suggestions.

Hello, here is a demo.

https://stackblitz.com/edit/vitejs-vite-v38snj?file=src/NumberField.jsx

**Comment by snowystinger** [comment]
Hey, apologies for the delay, I have not forgotten about this, but I have not had time to look into it yet. I'm hoping to get back to it within the week.

**Comment by LFDanLu** [comment]
@andelkocvjetkovic Apologies for the delay, I've taken a look at the your demo NumberField. So your current implementation retains the `step` behavior of enforcing that a user input value is a multiple of the provided `step`. However, you've overridden the buttons onPressStart so that the current value is incremented and decremented by the `customStep` value rather than by the `step`. Is that a fair summary? As an aside, would the above implementation also want the arrow keys to increment/decrement by `customStep` as well?

Just to clarify, which is the desired end behavior:
1.  Being able to manually input any value into the field and retain it after the blur, regardless of the `step` provided?
2. Being able to provide a value to the increment/decrement actions that is different from the `step` provided?
3. Both 1 and 2?

I _think_ you want 2, but lemme know if I'm wrong in my understanding here

**Comment by andelkocvjetkovic** [comment]
@LFDanLu, you are right, my implementation is not working correctly for arrow keys.
I'm not sure if is possible to accomplish 2. without 1. .
My case is, I wanna be able to input a value (ex. 225.00) but increment/decrement it by 100 without snap to(325.00 or 25.00) on inc/dec action.

**Comment by LFDanLu** [comment]
What step value would it snap to (I assume to the nearest .01 value)? In this case the step value would only be enforced on blur and a separate `customStep` value controls how much the inc/dec buttons + up/down arrow keys change the value correct? If so this feels like a pretty large departure from the current behavior/native, lemme ask the team how they feel about it

**Comment by LFDanLu** [comment]
@andelkocvjetkovic The team was hesitant about trying to add this to the NumberField hooks due to current in flight priorities. Some ideas that I had that you could perhaps try:
- overwrite the `increment` and `decrement` functions returned by useNumberFieldState in such a way that they accommodate your `customStep`. I haven't tried this myself so I'm unsure of the feasibility or scope of work required.
- Introduce separate increment/decrement button that when pressed change the value of your numberfield by +/- 100. Maybe not the cleanest design but I've seen numberfields with accompanying +10/+100/etc buttons in other places

**Comment by yihuiliao** [comment]
Closing since we do not plan to export `setNumberValue` and `safeNextStep` and there seems to be a workaround.
